### PR TITLE
Regex search for Projects and Documents

### PR DESCRIPTION
**Pull request by dhlevi**
Added a regex search in addition to our text search. Regex search will only be triggered if you include a keyword, but do not wrap it in quotes. It will filter out the text searches to literal matches from the words provided in the keyword string. I've also added score back in to the project search aggregate. Results with a score of less then 5000 will be omitted from the results.

Most Public and Admin pages now provide default sort params, so to get the requested behaviour of the closest matches bubbling to the top, those pages will have to be updated so a new keyword search resets the default sort to -score (separate PRs will be made for those in public and admin).

Update: Also added support for a fuzzy n-gram text search in addition to the regex search. Fuzzy text search can be used on document and project searches. Using fuzzy search will disable regex search.

See ticket [EE-1021](https://bcmines.atlassian.net/browse/EE-1021)

### CODE REVIEW

File: api/aggregators/searchAggregator.js [hunk 1]
```diff
@@ -93,6 +94,35 @@ exports.createMatchAggr = async (schemaName, projectId, keywords, caseSensitive,
...
+    let terms = decodedKeywords.split(' ');
+    let searchTerm = terms.join('|');
+
+    // By default, if we're doing a keyword search exclude
+    // any values that have a score less then 5000.
+    let regexMatch = { $match: { score: { $gt: 5000 } } };
+
+    let regex = { $regex:'(?:^|(?<= ))(' + searchTerm + ')(?:(?= )|$)', $options:'i' };
+
+    if (schemaName === constants.PROJECT) {
```
Comment (KitArmstrong) [inline api/aggregators/searchAggregator.js:114]: Should this live in this function, or should this function only create the Regex and not care about where it is being used?

File: api/controllers/search.js [hunk 2]
```diff
@@ -86,18 +86,22 @@ const searchCollection = async function (roles, keywords, schemaName, pageNum, p
...
   if (!schemaAggregation) {
-    aggregation = [...matchAggregation, ...sortingPagingAggr];
+    aggregation = [...matchAggregation, ...keywordRegexFilter, ...sortingPagingAggr];
   } else {
-    aggregation = [...matchAggregation, ...schemaAggregation, ...sortingPagingAggr];
-
+    aggregation = [...matchAggregation, ...schemaAggregation, ...keywordRegexFilter, ...sortingPagingAggr];
   }
 
+  console.log(JSON.stringify(aggregation));
```
Comment (KitArmstrong) [inline api/controllers/search.js:103]: Can be removed.

File: api/helpers/minio.js [hunk 3]
```diff
@@ -9,11 +9,11 @@ var _ = require('lodash');
  * The Minio client which facilitates the connection to Minio, and through which all calls should be made.
  */
 var minioClient = new minio.Client({
-  endPoint: process.env.MINIO_HOST,
-  port: 443,
-  useSSL: true,
-  accessKey: process.env.MINIO_ACCESS_KEY,
-  secretKey: process.env.MINIO_SECRET_KEY
+  endPoint: 'localhost',
```
Comment (KitArmstrong) [inline api/helpers/minio.js:12]: Reverted

File: api/controllers/search.js (hunk 2 above)
Comment (dhlevi) [reply api/controllers/search.js:103]: Whoops, checked in debug code. Will revert

File: api/aggregators/searchAggregator.js (hunk 1 above)
Comment (dhlevi) [reply api/aggregators/searchAggregator.js:114]: Thought about that. I guess I'm fairly indifference. My thought was other places might want specific regex functions for their tasks, so we'd wind up writing it again anyway

**Comment by dhlevi** [comment]
Nope, performance didn't seem to be affected. Even document search was about the same as before. Likely we could get away without the regex at all, but it definitely cleans up the results to make more sense.

It would be nice to include fuzzy searching, but that's not a default function in mongodb's text search (they used to recommend elastic for that, but now they have Atlas, which we could look in to)

**Comment by dhlevi** [comment]
@KitArmstrong Re-review if possible (or I can pass it on to someone else). I've added an additional option to support n-gram based fuzzy text searches.

**Pull request merged**